rust/headless-client/src/windows.rs

```diff
@@ -5,10 +5,211 @@
 
 use anyhow::Result;
-use std::path::{Path, PathBuf};
+use std::{
+    cmp::Ordering,
+    io,
+    mem::MaybeUninit,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+    path::{Path, PathBuf},
+    ptr::null,
+};
 
-pub(crate) use socket_factory::tcp as tcp_socket_factory;
-pub(crate) use socket_factory::udp as udp_socket_factory;
+use windows::Win32::NetworkManagement::{IpHelper::GetAdaptersAddresses, Ndis::NET_LUID_LH};
+use windows::Win32::Networking::WinSock::SOCKADDR_INET;
+use windows::Win32::{
+    NetworkManagement::IpHelper::{
+        GetBestRoute2, GET_ADAPTERS_ADDRESSES_FLAGS, IP_ADAPTER_ADDRESSES_LH, MIB_IPFORWARD_ROW2,
+    },
+    Networking::WinSock::{ADDRESS_FAMILY, AF_UNSPEC},
+};
+
+use firezone_bin_shared::TUNNEL_NAME;
+use socket_factory::{TcpSocket, UdpSocket};
+
+pub fn tcp_socket_factory(addr: &SocketAddr) -> io::Result<TcpSocket> {
+    let local =
+        get_best_non_tunnel_route(addr.ip())?.ok_or(io::Error::other("No route to host"))?;
+
+    let socket = socket_factory::tcp(addr)?;
+    socket.bind((local, 0).into())?; // To avoid routing loops, all TCP sockets are bound to "best" source IP.
+
+    Ok(socket)
+}
+
+pub fn udp_socket_factory(src_addr: &SocketAddr) -> io::Result<UdpSocket> {
+    let socket =
+        socket_factory::udp(src_addr)?.with_source_ip_resolver(Box::new(get_best_non_tunnel_route));
+
+    Ok(socket)
+}
+
+fn get_best_non_tunnel_route(dst: IpAddr) -> io::Result<Option<IpAddr>> {
+    let src = get_best_route_excluding_interface(dst, TUNNEL_NAME)
+        .ok_or(io::Error::other("No route to host"))?;
+
+    tracing::debug!(%src, %dst, "Resolved best route outside of tunnel interface");
+
+    Ok(Some(src))
+}
+
+struct Adapters {
+    _buffer: Vec<u8>,
+    next: *const IP_ADAPTER_ADDRESSES_LH,
+}
+
+impl Iterator for Adapters {
+    type Item = &'static IP_ADAPTER_ADDRESSES_LH;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // SAFETY: We expect windows to give us a valid linked list where each item fo the list is actually an IP_ADAPTER_ADDRESSES_LH.
+        // TODO: pointer alignment?
+        let adapter = unsafe { self.next.as_ref()? };
+
+        self.next = adapter.Next;
+
+        Some(adapter)
+    }
+}
+
+/// Finds the best route (i.e. source interface) for a given destination IP, excluding interfaces where the name matches the given filter.
+///
+/// To prevent routing loops on Windows, we need to explicitly set a source IP for all packets.
+/// Windows uses a computed metric per interface for routing.
+/// We implement the same logic here, with the addition of explicitly filtering out our TUN interface.
+///
+/// # Performance
+///
+/// This function performs multiple syscalls and is thus fairly expensive.
+/// It should **not** be called on a per-packet basis.
+/// Callers should instead cache the result until network interfaces change.
+fn get_best_route_excluding_interface(dst: IpAddr, filter: &str) -> Option<IpAddr> {
```

Comment:
I reckon lets actually get rid of the `filter` parameter and call this function `get_best_non_tunnel_route`.

Reply:
With the refactoring to make it memory safe, we can easily add the log statement in here.